rfcs/0009-defaults-and-overrides.md

```diff
@@ -0,0 +1,547 @@
+- `defaults`: a block of _default_ policy rules with further specification of a strategy (_atomic_ set of rules or individual rules to be _merged_ into lower policies), and optional conditions for applying the defaults down through the hierarchy
+- `overrides`: a block of _override_ policy rules with further specification of a strategy (_atomic_ set of rules or individual rules to be _merged_ into lower policies), and optional conditions for applying the overrides down through the hierarchy
+- the bare policy rules block without further qualification as a default or override set of rules – e.g. the `rules` field in a Kuadrant AuthPolicy, the `limits` field in a RateLimitPolicy.
+
+Typically, one will specify either `defaults` and/or `overrides`, or just the bare set of policy rules block without further qualification as neither defaults nor overrides.
+
+In case two or more of these fields are specified, they are processed in the following order:
+1. `defaults`
+2. bare set of policy rules without qualification (treated indistinctively as another block of defaults)
+3. `overrides`
+
+Supporting specifying the bare set of policy rules at the first level of the spec, as well and simultaneously along with `defaults` and `overrides` blocks, is a strategy that aims to provide:
+1. more natural usability, especially for those who write policies attached to the lowest level of the hierarchy supported; as well as
+2. backward compatibility for policies that did not support explicit D/O and later on have moved to doing so.
+
+### Direct Policies
+
+A policy that does not specify D/O fields (`defaults`, `overrides`) is a policy that _declares an intent_.
+
+When a policy of such kind targets an object whose kind is the lowest kind accepted by Kuadrant in the hierarchy of network resources, such policy is treated as a [_Direct Policy_](https://gateway-api.sigs.k8s.io/geps/gep-713/#direct-policy-attachment). When, otherwise, it is used to target "higher" objects, the policy is treated as an Inherited Policy that implicitly declares atomic _defaults_.
+
+### Inherited Policies
+
+A policy that specifies D/O fields (`defaults`, `overrides`) is a policy explicitly declared to _modify an intent_ (or a _proper Inherited Policy_.)
+
+There is no meaning in declaring a proper Inherited Policy that targets an object whose kind is the lowest kind accepted by Kuadrant in the hierarchy of network resources. The sets of rules specified in these policies affect indistinctively the targeted objects, whether the rules are qualified as `defaults` or as `overrides`.
+
+### Example of D/O-enabled Kuadrant policy
+
+**Example 1.** Atomic defaults
+
+```yaml
+kind: AuthPolicy
+metadata:
+  name: gw-policy
+spec:
+  targetRef:
+    kind: Gateway
+  defaults:
+    rules:
+      authentication:
+        "a": {…}
+      authorization:
+        "b": {…}
+    strategy: atomic
```

Comment:
I'd recommend moving to using lists of named subobjects instead of maps for the `rules` sections:
```yaml
kind: AuthPolicy
metadata:
  name: gw-policy
spec:
  targetRef:
    kind: Gateway
  defaults:
    rules:
      authentication:
        - name: "a"
          {...}
      authorization:
        - name: "b"
          {...}
    strategy: atomic
```

This is straight from the Kubernetes API conventions, [https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subo[…]preferred-over-maps](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps), and as it says "maintains the invariant that all JSON/YAML keys are fields in API objects"

kubernetes/kubernetes#2004 goes into some more detail of why this is probably better.

Reply:
We actually walked the opposite direction, [replacing lists with maps not too long ago](https://github.com/Kuadrant/authorino/pull/417). It was our (possibly naive?) attempt to prepare for supporting merges between policies that would acknowledge each of those named policy rules individually.

We knew that, with lists, a merge would have to pick either one entire list of rules or the other, but never mix and combine two lists. Something about that was recommended in [GEP-713](https://gateway-api.sigs.k8s.io/geps/gep-713/#merging-into-existing-spec-fields), I recon.

To support our use cases, we found no way but to treat those _user-defined values_ (names of rules within a policy) as keys, not as value. Using maps felt then the lesser of two evils perhaps.

Any suggestion that could help us solve the issue? Could we still merge lists picking elements of both?

Thanks for the references!

Reply:
The trick is to treat the list as what the apiserver calls a listMapType: although the objects are _stored_ as a list, all code that handles them treats them as a map with the `name` as the key. (This is actually how Conditions and Listeners work already).

Then the apiserver can use the Strategic merge type and handle a lot of this for you. 

In the CRD definitions, this is done with:
```go
	// +listType=map
	// +listMapKey=name
```
on the containing field (the `[]Type` field).

Doing things this way also gives a big advantage in that the list is ordered, so you can assume that the listed order is important if you want. A map is explicitly randomly ordered, so you'll need to do something else to sort the keys.

To be clear, this is a case of all the solutions having bad points. But personally, I think that the `listType=map` has the least onerous tradeoffs.

Reply:
Added as a future possibility as it's not a blocker.